src/Qir/Runtime/lib/Simulators/FullstateSimulator.cpp

```diff
@@ -408,7 +419,109 @@ namespace Quantum
+                }
+                return true;
+            };
+    }; // class CFullstateSimulator
+
+    void CFullstateSimulator::DumpMachineImpl(std::ostream& outStream)
+    {
+        outStream << "*********************" << std::endl;
+        this->GetStateTo((TDumpLocation)&outStream, dumpToLocationCallback);
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    void CFullstateSimulator::DumpRegisterImpl(std::ostream& outStream, const QirArray* qubits)
+    {
+        outStream << "*********************" << std::endl;
+        if(!this->GetRegisterTo((TDumpLocation)&outStream, dumpToLocationCallback, qubits))
+        {
+            outStream << "## Qubits were entangled with an external qubit. Cannot dump corresponding wave function. ##" << std::endl;
+        }
+        outStream << "*********************" << std::endl;
+        outStream.flush();
+    }
+
+    bool CFullstateSimulator::GetRegisterTo(TDumpLocation location, TDumpToLocationCallback callback, const QirArray* qubits)
+    {
+        // `qubits->buffer` points to the sequence of qubit ids each of type `void*` (logically `qubits->buffer` is of type `void**`).
+        // We need to pass to the native simulator the pointer to the sequence of `unsigned`.
+        std::vector<unsigned> unsQbIds(qubits->count);
+        size_t i = 0;
+        for (unsigned& qbId : unsQbIds)
+        {
+            qbId = (unsigned)(uintptr_t)(((void**)(qubits->buffer))[i]);
+            ++i;
+        }
+        static TDumpQubitsToLocationAPI dumpQubitsToLocation =
+            reinterpret_cast<TDumpQubitsToLocationAPI>(this->GetProc("DumpQubitsToLocation"));
+        return dumpQubitsToLocation(
+            this->simulatorId, (unsigned)(qubits->count), unsQbIds.data(), callback,
+            location);
+    }
+
+    void CFullstateSimulator::GetStateTo(TDumpLocation location, TDumpToLocationCallback callback)
+    {
+        static TDumpToLocationAPI dumpTo = reinterpret_cast<TDumpToLocationAPI>(this->GetProc("DumpToLocation"));
+        
+        dumpTo(this->simulatorId, callback, location);
+    }
+
+    std::ostream& CFullstateSimulator::GetOutStream(const void* location, std::ofstream& outFileStream)
+    {
+        // If the location is not nullptr and not empty string then dump to a file:
+        if((location != nullptr) &&
+            (((static_cast<const QirString *>(location))->str) != ""))
+        {
+            // Open the file for appending:
+            const std::string& filePath = (static_cast<const QirString *>(location))->str;
+
+            outFileStream.open(filePath, std::ofstream::out | std::ofstream::app);
+            if((outFileStream.rdstate() & std::ofstream::failbit) != 0)
+            {
+                QirString qstr {std::string("Failed to open dump file \"") +
+                                            filePath +
+                                            "\""};
+                quantum__rt__fail(&qstr);

```

Comment:
Similarly here, hitting a `fail` call isn't ideal for a diagnostic routine.

Reply:
Fixed.